lib/chef/resource/aws_route53_hosted_zone.rb

```diff
@@ -0,0 +1,220 @@
+
+          # we already checked for duplicate Chef RR resources in #get_record_sets_from_resource.
+          keyed_chef_resources = record_set_resources.reduce({}) { |coll, rs| (coll[rs.aws_key] ||= []) << rs; coll }
+          keyed_aws_objects    = aws_record_sets.reduce({})      { |coll, rs| coll[rs.aws_key] = rs; coll }
+
+          keyed_chef_resources.each do |key, chef_resource_ary|
+            chef_resource_ary.each do |chef_resource|
+              # RR already exists...
+              if keyed_aws_objects.has_key?(key)
+                # ... but we want to delete or update it.
+                action = chef_resource.action.first == :destroy ? DELETE : UPDATE
+                change_list << chef_resource.to_aws_change_struct(action)
+              # otherwise, RR does not already exist...
+              else
+                # using UPSERT instead of CREATE: it will likely DTRT for the user if our code has a bug.
+                change_list << chef_resource.to_aws_change_struct(UPSERT)
+              end
+            end
+          end
+
+          Chef::Log.debug("RecordSet changes: #{change_list.inspect}")
+          if change_list.size > 0
+            new_resource.driver.route53_client.change_resource_record_sets(hosted_zone_id: new_resource.aws_route53_zone_id,
+                                                                           change_batch: {
+                                                                             comment: RRS_COMMENT,
+                                                                             changes: change_list,
+                                                                             })
+          else
+            Chef::Log.info("All aws_route53_record_set resources up to date (nothing to do).")
+          end
+        end
+      rescue => ex
+        # the RecordSet change is transactional, but we must revert the comment manually.
+        if new_resource.comment != hosted_zone.config.comment
+          new_resource.driver.route53_client.update_hosted_zone_comment(id: hosted_zone.id, comment: hosted_zone.config.comment)
+        end
+        raise
+      end
+    end
+  end
+
+  def destroy_aws_object(hosted_zone)
+    converge_by "delete Route53 zone #{new_resource}" do
+      Chef::Log.info("Deleting all non-SOA/NS records for #{hosted_zone.name}")
+
+      rr_changes = hosted_zone.resource_record_sets.reject { |aws_rr|
+        %w{SOA NS}.include?(aws_rr.type)
+        }.map { |aws_rr|
+          {
+            action: DELETE,
+            resource_record_set: aws_rr.to_change_struct,
+          }
+        }
+
+      if rr_changes.size > 0
+        aws_struct = {
+          hosted_zone_id: hosted_zone.id,
+          change_batch: {
+            comment: "Purging RRs prior to deleting resource",
+            changes: rr_changes,
+          }
+        }
+
+        new_resource.driver.route53_client.change_resource_record_sets(aws_struct)
```

Comment:
The documentation says the status refers to whether the changes have propagated to all of AWS's DNS servers, but the AWS object creation is instantaneous, so no polling needed.